### PR TITLE
 Improve bitmap index matching to better leverage the multi-key indexes

### PR DESCRIPTION
**Pull request by dhanashreek89**
This PR consists of following two major commits, the rest commits add new tests, modify existing ones and add a few utility functions in `CPredicateUtils` needed for the new algorithm.

- Improve bitmap index matching to better leverage the multi-key indexes

Do not break down a predicate that contains conjnucts with children of the form:
"Ident Cmp Const" or "Ident Cmp Const-array" or "Boolean Ident" or "NOT (Boolean Ident)".
Instead treat the predicate as a whole while finding a suitable index match.  Also, for such predicates, perform the lookup again for the residual to minimize the number of final residuals and to leverage available indexes on them.
If there are multiple index options available for the given predicate then an Index covering maximum predicate columns is selected.

Example, for schema `t(a, b, c, f, e)` indexes: `i1 (a)`,  `i2 (e)`,  `i3 (a, b)` and `i4 (b, f, a)`
and Predicate: `(a = 1) AND (b = 2) AND (e = 3) AND (c = 4)`

Previously, we separated the children of AND and performed the lookup independently for each of them; also the lookup stopped at first matching index. Hence if there were multiple index matches available for a child, it would pick the first one in the OID order. For instance, for `(a =1)` if `i1` was first in the list then it would get picked without considering the rest of the indexes which could be potential matches too. The final index plan would look like:
```
BitmapTableScan
   Recheck: (a = 1) AND (b = 2) AND (e = 3)
   Filter: c = 4
      |_
         BitmapAnd
	    |_
              BitmapIndexScan on I1
	         Index Cond: a = 1
	    |_
	      BitmapIndexScan on I3
	         Index Cond: b = 2
            |_
	      BitmapIndexScan on I2
	         Index Cond: e = 3
```
After this commit, the lookup will go through all the available indexes to pick an optimal index. It will produce:
```
BitmapTableScan
   Recheck: (a = 1) AND (b = 2) AND (e = 3)
   Filter: c = 4
      |_
         BitmapAnd
	    |_
	      BitmapIndexScan on I3
	         Index Cond: (a = 1) AND (b = 2)
            |_
	      BitmapIndexScan on I2
	         Index Cond: e = 3
```
If predicate is a disjunct tree, then algorithm behaves same as before wherein the disjunct tree is broken down and index lookup is performed on individual children. Additionally in case of join2Index transforms, the matching algorithm remains same.

In nutshell, the new and improved index matching algorithm is only specialized for predicates of the form with a schema containing multi-key indexes (which is very common) :
`SELECT * FROM t1 WHERE c = 10 AND b = 40 AND g = 35 AND e = 20 AND h = 50;`

- Do not select the index if the leading key of a multi key index is not covered by the predicate,

because such index scan will be expensive. If a multi-key index has been selected to be a potential match for given predicate, then make sure that first key of the index is present in the
predicate. For example, with schema: `t(a, b , c, d)`, index `i(b ,a, c)` and predicate: `c = 10`,

No index path will be chosen since the given predicate does not cover first key of the index i. However if the predicate is (b = 1 and c = 10) then this index can be selected.

Ideally, two alternatives should be generated in this case,
1. Table scan with `c = 10` as filter
2. Index scan with index `i` (bad index choice)
And the cost model should penalize the bad index scans and make the decision. Currently, our cost model does not cost index scans used for predicates that skip the first index keys, differently, than an index scan used for a predicate that covers the first key of the multi-key index. With cost model equipped to handle such index paths, we can generate such paths and leave it up to the costing. But for now, disallow them.

Following is the simple setup that shows the performance improvement after picking optimal indexes:
```
create table t1 (a int, b int, c int, d int) with (appendonly = true, orientation=column) distributed by (a);
create index t1_b_d on t1 using btree (b, d);
create index t1_c_d on t1 using btree (c,d);
insert into t1 select i, i%100000, i%100000, i%100000 from generate_series (1, 1000000) i;
```

Before, the following query ran for `~14 ms` :
```
explain analyze select * from t1 where c = 100 and d = 10;
                                                      QUERY PLAN
----------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.00 rows=1 width=16)
   Rows out:  0 rows at destination with 14 ms to end.
   ->  Bitmap Table Scan on t1  (cost=0.00..431.00 rows=1 width=16)
         Recheck Cond: c = 100 AND d = 10
         Rows out:  0 rows (seg0) with 14 ms to end.
         ->  BitmapAnd  (cost=0.00..0.00 rows=0 width=0)
               Bitmaps out:  0 (seg0) with 14 ms to end, start offset by 0.606 ms.
               ->  Bitmap Index Scan on t1_c_d  (cost=0.00..0.00 rows=0 width=0)
                     Index Cond: c = 100
                     Bitmaps out:  Avg 1.0 x 3 workers.  Max 1 (seg0) with 0.021 ms to end, start offset by 0.599 ms.
                     Work_mem used:  27K bytes avg, 33K bytes max (seg1).
               ->  Bitmap Index Scan on t1_b_d  (cost=0.00..0.00 rows=0 width=0)
                     Index Cond: d = 10
                     Bitmaps out:  Avg 1.0 x 3 workers.  Max 1 (seg0) with 14 ms to end, start offset by 0.621 ms.
                     Work_mem used:  27K bytes avg, 41K bytes max (seg1).
```

After this PR, ORCA will chose optimal index covering both `c` and `d` and run time is `~0.5ms`:
```
explain analyze select * from t1 where c = 100 and d = 10;
                                                   QUERY PLAN
----------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.00 rows=1 width=16)
   Rows out:  0 rows at destination with 0.484 ms to end.
   ->  Bitmap Table Scan on t1  (cost=0.00..431.00 rows=1 width=16)
         Recheck Cond: c = 100 AND d = 10
         Rows out:  0 rows (seg0) with 0.166 ms to end.
         ->  Bitmap Index Scan on t1_c_d  (cost=0.00..0.00 rows=0 width=0)
               Index Cond: c = 100 AND d = 10
               Bitmaps out:  Avg 1.0 x 3 workers.  Max 1 (seg0) with 0.013 ms to end, start offset by 0.560 ms.
```

### CODE REVIEW

File: libgpopt/include/gpopt/operators/CPredicateUtils.h [hunk 1]
```diff
@@ -565,6 +576,11 @@ namespace gpopt
 			static
 			BOOL FConvertToCNF(CExpression *pexpr, 	CExpression *pexprOuter, CExpression *pexprInner);
```
Comment (melanieplageman) [inline libgpopt/include/gpopt/operators/CPredicateUtils.h]: Would you add a comment about what CNF is?

File: libgpopt/src/operators/CPredicateUtils.cpp [hunk 2]
```diff
@@ -2708,5 +2786,38 @@ CPredicateUtils::FCollapsibleChildUnionUnionAll
 	return (pexprChild->Pop()->Eopid() == pexpr->Pop()->Eopid());
 }
 
+// check if the expression is a conjunct with children of the form:
+// <ident/casted ident> op <const/casted const> or
+// <ident/casted ident> op <const array/casted const array>
+// <boolean ident> or
+// NOT <boolean ident>
+BOOL
+CPredicateUtils::FAndWithScalarIdentToScalarConstChildren
```
Comment (melanieplageman) [inline libgpopt/src/operators/CPredicateUtils.cpp]: The syntax of the comment is a little hard to parse as the reader. It would be great if you could give more detail and maybe an example.

File: libgpopt/src/operators/CPredicateUtils.cpp [hunk 3]
```diff
@@ -1227,22 +1239,58 @@ CPredicateUtils::FIdentINDFConstIgnoreCast
 	return FIdentCompareConstIgnoreCast((*pexpr)[0], COperator::EopScalarIsDistinctFrom);
 }
 
+// is the given expression a comparison between a scalar ident under a scalar cast and a constant array
+// +--CScalarArrayCmp Any (=)
```
Comment (melanieplageman) [inline libgpopt/src/operators/CPredicateUtils.cpp]: This comment doesn't look right--there is no constant array here

File: data/dxl/minidump/PredicateWithConjunctsAndDisjuncts.mdp [hunk 4]
```diff
@@ -0,0 +1,8013 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+CREATE TABLE t1 (a int, b int, c int, d int, e int, f int, g int, h int) WITH (appendonly = true, orientation=column) distributed by (a);
+CREATE INDEX t1_b on t1 USING btree(b);
+CREATE INDEX t1_b_c_d on t1 USING btree(b,c,d);
+CREATE INDEX t1_g_h on t1 USING btree(g,h);
+INSERT INTO t1 SELECT i, i%100000, i%100000, i%100000, i%100000, i%100000, i%100000, i%100000 FROM generate_series (1, 1000000) i;
+EXPLAIN SELECT * FROM t1 WHERE b = 10 and e = 10 and (c = 40 and d = 35) or (g = 50 and h = 60);
```
Comment (melanieplageman) [inline data/dxl/minidump/PredicateWithConjunctsAndDisjuncts.mdp]: It would be helpful to put in the comment either the expected explain plan or one line with the expected behavior. Specifically, what our behavior under test is and what our assumptions are. btree on AO table test for a PR about bitmap indexes is confusing unless you know the context. Also, if we are only checking correctness of indexes chosen and that is why we don't analyze the table because costing would make this test not work (for example)

File: libgpopt/src/xforms/CXformUtils.cpp [hunk 5]
```diff
@@ -2964,25 +2964,6 @@ CXformUtils::PexprScalarBitmapBoolOp
 		
 	const ULONG ulPredicates = pdrgpexpr->UlLength();
 	
-	if (1 == ulPredicates)
-	{
-		return PexprBitmap
```
Comment (melanieplageman) [inline libgpopt/src/xforms/CXformUtils.cpp]: Could you separate the bitmap-related utils functions out of this file into a different file specific to bitmap. It would help improve readability and reduce file size to make it more manageable.

File: data/dxl/minidump/PredicateWithConjunctsAndDisjuncts.mdp (hunk 4 above)
Comment (dhanashreek89) [reply data/dxl/minidump/PredicateWithConjunctsAndDisjuncts.mdp]: >> It would be helpful to put in the comment either the expected explain plan or one line with the expected behavior. Specifically, what our behavior under test is and what our assumptions are. btree on AO table test for a PR about bitmap indexes is confusing unless you know the context.

Done

>> Also, if we are only checking correctness of indexes chosen and that is why we don't analyze the table because costing would make this test not work 

No, the minidump contains valid statistics on the table. The statistics were collected during the `INSERT` statement since the autostats was ON.
The intent is to see a final plan with correct index scans generated, so statistics are needed.

File: data/dxl/minidump/MultipleIndexPredicate.mdp [hunk 6]
```diff
@@ -0,0 +1,1711 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
+  <dxl:Comment><![CDATA[
+	CREATE TABLE foo (
```
Comment (oarap) [inline data/dxl/minidump/MultipleIndexPredicate.mdp]: why do we need this many columns?

File: libgpopt/src/operators/CPredicateUtils.cpp [hunk 7]
```diff
@@ -1185,30 +1183,44 @@ CPredicateUtils::FIdentCompareConstIgnoreCast
...
 	}
 
-	// cast(col) IDF cast(constant)
+	// cast(col) cmp cast(constant)
 	if (CScalarIdent::FCastedScId(pexprLeft) && CScalarConst::FCastedConst(pexprRight))
 	{
 		return true;
 	}
 
+	// col cmp constant-array
```
Comment (oarap) [inline libgpopt/src/operators/CPredicateUtils.cpp]: This part should be a separate function. It does not match with the description or naming of this function.

File: libgpopt/src/operators/CPredicateUtils.cpp [hunk 8]
```diff
@@ -1227,22 +1239,58 @@ CPredicateUtils::FIdentINDFConstIgnoreCast
...
+		return false;
+	}
+
+	CExpression *pexprLeft = (*pexpr)[0];
+	CExpression *pexprRight = (*pexpr)[1];
+
+	if (CCastUtils::FBinaryCoercibleCast(pexprLeft) &&
+		 CUtils::FScalarIdent((*pexprLeft)[0]))
+	{
+		return CUtils::FScalarConst(pexprRight);
```
Comment (oarap) [inline libgpopt/src/operators/CPredicateUtils.cpp]: Isn't this supposed to be `FScalarArray`?

File: libgpopt/src/xforms/CXformUtils.cpp [hunk 9]
```diff
@@ -3281,24 +3214,42 @@ CXformUtils::PexprBitmap
...
+	// for rest of the expression types, such as <ident> op <ident>,
+	// disjuncts or a conjunct containing any of these, look for index
+	// access paths in its children
+	if (!CPredicateUtils::FAndWithScalarIdentToScalarConstChildren(pexprPred))
 	{
+		(*retryIndexLookupWithResidual) = false;
+
 		// first, check if it is an index lookup predicate
+		// if the predicate is of the form <ident> op <ident>
 		CExpression *pexprBitmapForIndexLookup = PexprBitmapForIndexLookup
```
Comment (hardikar) [inline libgpopt/src/xforms/CXformUtils.cpp]: PexprBitmapForIndexLookup doesn't really capture the intent of "check if it is an index lookup predicate if the predicate is of the form <ident> op <ident>".
Let's add some more comments here; and/or rename that method as well?

comments on the method:
```
//	@function:
//		CXformUtils::PexprBitmapForIndexLookup
//
//	@doc:
//		Construct a bitmap index path expression for the given predicate coming
//		from a condition with outer references that could potentially become
//		an index lookup
```

File: libgpopt/src/xforms/CXformUtils.cpp [hunk 10]
```diff
@@ -3331,13 +3283,23 @@ CXformUtils::PexprBitmap
...
 											ppexprResidual
 											);
-		if (NULL != pexprBitmapFromChildren)
+
+		// if no index path was constructed for this predicate, return
+		// it as residual.
+		// Example, with schema: t(a, b, c, d), index i1(a,b)
+		// and predicate: (a = 3) OR (c = 4 AND d =5)
+		// no index path will be found for (c = 4 AND d =5)
+		// in which case the entire disjunct will become a residual
```
Comment (hardikar) [inline libgpopt/src/xforms/CXformUtils.cpp]: How does this example work when its inside the `if !FAndWithScalarIdentToScalarConstChildren` block?

File: libgpopt/src/xforms/CXformUtils.cpp [hunk 11]
```diff
@@ -3439,22 +3394,58 @@ CXformUtils::PexprBitmapForSelectCondition
...
-							GPOS_NEW(pmp) CScalarBitmapIndexProbe(pmp, pindexdesc, pmdindex->PmdidItemType()),
-							pdrgpexprIndex
-							);
+			// if this index covers more columns than a previously found index, replace best index match
+			ULONG ulResidualLength = pdrgpexprResidual->UlLength();
+			if (minResidual > ulResidualLength)
+			{
+				CRefCount::SafeRelease((*ppexprResidual));
+				pdrgpexprResidual->AddRef();
+				(*ppexprResidual) = CPredicateUtils::PexprConjDisj(pmp, pdrgpexprResidual, true);
```
Comment (hardikar) [inline libgpopt/src/xforms/CXformUtils.cpp]: can we add a comment next to true /* fConjunction */ or whatever the param is.

File: libgpopt/src/xforms/CXformUtils.cpp [hunk 12]
```diff
@@ -3439,22 +3394,58 @@ CXformUtils::PexprBitmapForSelectCondition
...
-			pmdindex->PmdidItemType()->AddRef();
-			return 	GPOS_NEW(pmp) CExpression
-							(
-							pmp,
-							GPOS_NEW(pmp) CScalarBitmapIndexProbe(pmp, pindexdesc, pmdindex->PmdidItemType()),
-							pdrgpexprIndex
-							);
+			// if this index covers more columns than a previously found index, replace best index match
+			ULONG ulResidualLength = pdrgpexprResidual->UlLength();
+			if (minResidual > ulResidualLength)
```
Comment (hardikar) [inline libgpopt/src/xforms/CXformUtils.cpp]: Ah we're trying to minimize the ulResidualLength per pexprPred! Can we add a comment to that effect?

File: libgpopt/src/xforms/CXformUtils.cpp [hunk 13]
```diff
@@ -3439,22 +3394,58 @@ CXformUtils::PexprBitmapForSelectCondition
...
+			{
+				CRefCount::SafeRelease((*ppexprResidual));
+				pdrgpexprResidual->AddRef();
+				(*ppexprResidual) = CPredicateUtils::PexprConjDisj(pmp, pdrgpexprResidual, true);
+
+				// discard the trivial filter that gets generated when index covers all the columns,
+				// in which case the residual is const true
+				if (CUtils::FScalarConstTrue((*ppexprResidual)))
```
Comment (hardikar) [inline libgpopt/src/xforms/CXformUtils.cpp]: Could this go outside the if block?

File: libgpopt/src/xforms/CXformUtils.cpp [hunk 14]
```diff
@@ -3792,7 +3875,7 @@ CXformUtils::PexprBitmapTableGet
 				pdrgpcrOutput,
 				pcrsOuterRefs,
 				pcrsReqd,
-				true, // fConjunction
+				fConjunction,
```
Comment (hardikar) [inline libgpopt/src/xforms/CXformUtils.cpp]: Since fConjunction depends on pexprScalar, we can probably remove this param :)

File: libgpopt/src/xforms/CXformUtils.cpp [hunk 15]
```diff
@@ -3570,65 +3561,152 @@ CXformUtils::CreateBitmapIndexProbeOps
...
 		}
+
+		// cleanup
+		CRefCount::SafeRelease(pdrgpexprBitmapTemp);
+		CRefCount::SafeRelease(pdrgpexprRecheckTemp);
 	}
 }
 
+void
+CXformUtils::JoinBitmapIndexProbes
```
Comment (hardikar) [inline libgpopt/src/xforms/CXformUtils.cpp]: Needs comments. A small one is enough.

File: libgpopt/src/xforms/CXformUtils.cpp (hunk 12 above)
Comment (hardikar) [reply libgpopt/src/xforms/CXformUtils.cpp]: And if there are multiple indexes that are equally viable in number of columns, will we pick the first one?

File: libgpopt/src/xforms/CXformUtils.cpp [hunk 16]
```diff
@@ -3570,65 +3561,152 @@ CXformUtils::CreateBitmapIndexProbeOps
...
-				pexprRecheck->Release();
+				GPOS_ASSERT(NULL != pexprRecheck);
+
+				// if the original predicate is a conjuct and if the matched index has already
+				// been selected previously, then merge them
+				// Example, for the schema: t1(c1), t2(c2) and index i1(t2.c2)
+				// and query: select * from t1, t2 where t1.c1 = t2.c2 and t2.c2 = 10;
+				// in case of Join2Index apply xform, i1 will be selected for both the predicates,
+				// hence can be merged
+				BOOL fAddedToPredicate = fConjunction  && FMergeWithPreviousBitmapIndexProbe
```
Comment (hardikar) [inline libgpopt/src/xforms/CXformUtils.cpp]: Did you consider to push this deduplication step to the very end instead of doing this per expr? I'm not sure of the trade of here - maybe its better right here.

File: libgpopt/src/xforms/CXformUtils.cpp [hunk 17]
```diff
@@ -3423,6 +3436,7 @@ CXformUtils::PexprBitmapForSelectCondition
 
 			pdrgpexprIndex->Release();
 			pdrgpexprResidual->Release();
+			indexColumns->Release();
```
Comment (hardikar) [inline libgpopt/src/xforms/CXformUtils.cpp]: Can we release this closer to where it's no longer used?

File: libgpopt/src/operators/CPredicateUtils.cpp (hunk 8 above)
Comment (dhanashreek89) [reply libgpopt/src/operators/CPredicateUtils.cpp]: No. This utility is to check if predicate is array comparison with a single element in array, in which case we have a single `ScalarConst` and not `ScalarConstArray`.

File: libgpopt/src/xforms/CXformUtils.cpp (hunk 9 above)
Comment (hardikar) [reply libgpopt/src/xforms/CXformUtils.cpp]: 1. Add comments about which transforms lead us here.
2. Is ident op ident possible with Select2Index transform?

File: libgpopt/src/xforms/CXformUtils.cpp [hunk 18]
```diff
@@ -3439,22 +3394,58 @@ CXformUtils::PexprBitmapForSelectCondition
...
+				CRefCount::SafeRelease((*ppexprResidual));
+				pdrgpexprResidual->AddRef();
+				(*ppexprResidual) = CPredicateUtils::PexprConjDisj(pmp, pdrgpexprResidual, true);
+
+				// discard the trivial filter that gets generated when index covers all the columns,
+				// in which case the residual is const true
+				if (CUtils::FScalarConstTrue((*ppexprResidual)))
+				{
+					(*ppexprResidual)->Release();
+					(*ppexprResidual) = NULL;
```
Comment (hardikar) [inline libgpopt/src/xforms/CXformUtils.cpp]: We can possibly break from the loop here.

File: libgpopt/src/xforms/CXformUtils.cpp [hunk 19]
```diff
@@ -2935,25 +2935,6 @@ CXformUtils::PexprScalarBitmapBoolOp
 		
 	const ULONG ulPredicates = pdrgpexpr->UlLength();
 	
-	if (1 == ulPredicates)
```
[the remark this answers is not in the document]
Comment (dhanashreek89) [reply libgpopt/src/xforms/CXformUtils.cpp]: Yes we no longer need to special case simple predicates, this will be handled in `CreateBitmapIndexProbeOps()`.

File: libgpopt/src/xforms/CXformUtils.cpp (hunk 12 above)
Comment (dhanashreek89) [reply libgpopt/src/xforms/CXformUtils.cpp]: >> And if there are multiple indexes that are equally viable in number of columns, will we pick the first one?

Yes

File: libgpopt/src/xforms/CXformUtils.cpp (hunk 15 above)
Comment (dhanashreek89) [reply libgpopt/src/xforms/CXformUtils.cpp]: comment added.

File: libgpopt/src/xforms/CXformUtils.cpp (hunk 14 above)
Comment (dhanashreek89) [reply libgpopt/src/xforms/CXformUtils.cpp]: Unfortunately we can not. The `fConjunction` can not be always derived from the`pexprScalar` in `PexprScalarBitmapBoolOp()`.
`fConjunction` tells us how the individual predicate expressions in `pdrgpexpr` are related. At the start of the index lookup, `pdrgpexpr` is generated from `pexprScalar` in which case `fConjunction` can be derived from the scalar expression.
However when `PexprScalarBitmapBoolOp()` is used in recursive context by `PexprBitmapFromChildren()`, `pdrgpexpr` is derived from children of `pexprScalar` hence it is incorrect to derive `fConjunction` from the original scalar, and thus needs to be sent explicitly.

File: libgpopt/src/xforms/CXformUtils.cpp (hunk 5 above)
Comment (dhanashreek89) [reply libgpopt/src/xforms/CXformUtils.cpp]: Good point!  There are also B-tree index related functions in this file which should be moved too. I plan to do this separately to keep the scope of this PR minimal.
